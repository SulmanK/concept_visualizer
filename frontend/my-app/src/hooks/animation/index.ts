export { default as useAnimatedMount } from "./useAnimatedMount";
export { default as useAnimatedValue } from "./useAnimatedValue";
export { default as usePrefersReducedMotion } from "./usePrefersReducedMotion";

// Re-export types
export type { AnimationState, AnimatedMountOptions } from "./useAnimatedMount";
export type { AnimatedValueOptions } from "./useAnimatedValue";
