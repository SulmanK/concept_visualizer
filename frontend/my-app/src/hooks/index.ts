// API Hooks
export { useConceptDetail, useRecentConcepts } from "./useConceptQueries";
export {
  useGenerateConceptMutation,
  useRefineConceptMutation,
} from "./useConceptMutations";
export {
  useRateLimitsQuery,
  useOptimisticRateLimitUpdate,
} from "./useRateLimitsQuery";
export { useExportImageMutation, downloadBlob } from "./useExportImageMutation";
export { useTaskStatusQuery, useTaskCancelMutation } from "./useTaskQueries";
export { useTaskSubscription } from "./useTaskSubscription";
export { useSessionSyncMutation } from "./useSessionQuery";
export { useConfigQuery } from "./useConfigQuery";

// Auth Hooks
export {
  useAuth,
  useAuthUser,
  useUserId,
  useIsAnonymous,
  useAuthIsLoading,
} from "./useAuth";

// Rate Limit Hooks
export {
  useRateLimitsData,
  useRateLimitsLoading,
  useRateLimitsError,
  useRateLimitsRefetch,
  useRateLimitsDecrement,
  useRateLimitContext,
  useDecrementRateLimit,
} from "./useRateLimits";

// Task Hooks
export {
  useTaskContext,
  useHasActiveTask,
  useIsTaskProcessing,
  useIsTaskPending,
  useIsTaskCompleted,
  useIsTaskFailed,
  useLatestResultId,
  useTaskInitiating,
  useActiveTaskId,
  useOnTaskCleared,
  useActiveTaskData,
  useClearActiveTask,
} from "./useTask";

// Error and Loading Hooks
export { useToast } from "./useToast";
export { ToastProvider } from "./toast/ToastContext";
export { useErrorHandling } from "./useErrorHandling";
export { useNetworkStatus } from "./useNetworkStatus";

// Animation Hooks
export {
  useAnimatedMount,
  useAnimatedValue,
  usePrefersReducedMotion,
} from "./animation";

// Animation Types (re-exported)
export type {
  AnimationState,
  AnimatedMountOptions,
  AnimatedValueOptions,
} from "./animation";
