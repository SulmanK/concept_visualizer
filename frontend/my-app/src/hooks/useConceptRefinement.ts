/**
 * Hook for refining visual concepts via the API.
 */

import { useState, useCallback } from 'react';
import { useApi } from './useApi';
import { 
  GenerationResponse,
  RefinementRequest,
  FormStatus
} from '../types';

export interface ConceptRefinementState {
  status: FormStatus;
  result: GenerationResponse | null;
  error: string | null;
}

/**
 * Hook for refining existing visual concepts
 */
export function useConceptRefinement() {
  const { post, loading, error, clearError } = useApi();
  const [refinementState, setRefinementState] = useState<ConceptRefinementState>({
    status: 'idle',
    result: null,
    error: null,
  });

  /**
   * Refine an existing concept based on provided instructions
   */
  const refineConcept = useCallback(async (
    originalImageUrl: string,
    refinementPrompt: string,
    logoDescription?: string,
    themeDescription?: string,
    preserveAspects: string[] = []
  ) => {
    // Validate inputs
    if (!originalImageUrl || !refinementPrompt.trim()) {
      setRefinementState({
        status: 'error',
        result: null,
        error: 'Please provide both an image to refine and refinement instructions',
      });
      return;
    }

    try {
      setRefinementState({
        status: 'submitting',
        result: null,
        error: null,
      });

      const request: RefinementRequest = {
        originalImageUrl,
        refinementPrompt,
        logoDescription,
        themeDescription,
        preserveAspects
      };

      const response = await post<GenerationResponse>('/concepts/refine', request);

      if (response.error) {
        setRefinementState({
          status: 'error',
          result: null,
          error: response.error.message,
        });
        return;
      }

      if (response.data) {
        setRefinementState({
          status: 'success',
          result: response.data,
          error: null,
        });
      }
    } catch (err) {
      setRefinementState({
        status: 'error',
        result: null,
        error: err instanceof Error ? err.message : 'Failed to refine concept',
      });
    }
  }, [post]);

  /**
   * Reset the refinement state
   */
  const resetRefinement = useCallback(() => {
    setRefinementState({
      status: 'idle',
      result: null,
      error: null,
    });
    clearError();
  }, [clearError]);

  return {
    refineConcept,
    resetRefinement,
    ...refinementState,
    isLoading: refinementState.status === 'submitting',
  };
} 