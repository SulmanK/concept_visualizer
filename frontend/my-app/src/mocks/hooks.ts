import React from "react";
import { vi } from "vitest";

// =============================================================================
// Mock Auth Context
// =============================================================================

// Create a context for our mock auth
export const mockAuthContext = React.createContext({
  user: { id: "test-user-id", email: "test@example.com" },
  session: { user: { id: "test-user-id", email: "test@example.com" } },
  isLoading: false,
  isAuthenticated: true,
  isAnonymous: false,
  signOut: vi.fn(),
  linkEmail: vi.fn(),
  refreshSession: vi.fn(),
  error: null,
});

// Mock auth context values
export const mockAuthContextValue = {
  user: { id: "test-user-id", email: "test@example.com" },
  session: { user: { id: "test-user-id", email: "test@example.com" } },
  isLoading: false,
  isAuthenticated: true,
  isAnonymous: false,
  signOut: vi.fn(),
  linkEmail: vi.fn(),
  refreshSession: vi.fn(),
  error: null,
};

// Create the auth hooks that match the ones in the real AuthContext
export const useAuth = () => React.useContext(mockAuthContext);
export const useAuthUser = () => React.useContext(mockAuthContext).user;
export const useUserId = () => React.useContext(mockAuthContext).user?.id;
export const useIsAnonymous = () =>
  React.useContext(mockAuthContext).isAnonymous;
export const useAuthIsLoading = () =>
  React.useContext(mockAuthContext).isLoading;

// =============================================================================
// Mock Toast Context
// =============================================================================

// Create a mock Toast context
export const mockToastContext = React.createContext({
  showToast: vi.fn(),
  showSuccess: vi.fn(),
  showError: vi.fn(),
  showWarning: vi.fn(),
  showInfo: vi.fn(),
  dismissToast: vi.fn(),
  dismissAll: vi.fn(),
});

// Hook for accessing toast context
export const useToast = () => React.useContext(mockToastContext);

// Mock toast context values
export const mockToastContextValue = {
  showToast: vi.fn(),
  showSuccess: vi.fn(),
  showError: vi.fn(),
  showWarning: vi.fn(),
  showInfo: vi.fn(),
  dismissToast: vi.fn(),
  dismissAll: vi.fn(),
};
