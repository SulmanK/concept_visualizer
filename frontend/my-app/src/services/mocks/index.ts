/**
 * Export all mock services and utilities for tests.
 * This file serves as the main entry point for importing all mock-related functionality.
 */

export * from "./mockApiService";
export * from "./mockHooks";
export * from "./testSetup";
