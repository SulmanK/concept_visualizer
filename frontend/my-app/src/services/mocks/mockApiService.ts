/**
 * Mock API service for testing purposes.
 * This service simulates responses from the backend API without making actual network calls.
 */

import {
  ApiResponse,
  ApiError,
  GenerationResponse,
  PromptRequest,
  RefinementRequest,
} from "../../types";

/**
 * Default mock responses for various API endpoints
 */
const mockResponses = {
  generateConcept: {
    success: {
      imageUrl: "https://example.com/mock-concept.png",
      colorPalette: {
        primary: "#4F46E5",
        secondary: "#818CF8",
        accent: "#C4B5FD",
        background: "#F5F3FF",
        text: "#1E1B4B",
      },
      generationId: "mock-gen-123",
      createdAt: new Date().toISOString(),
    },
    error: {
      status: 500,
      message: "Failed to generate concept",
    },
  },
  refineConcept: {
    success: {
      imageUrl: "https://example.com/mock-refined-concept.png",
      colorPalette: {
        primary: "#6366F1",
        secondary: "#A5B4FC",
        accent: "#E0E7FF",
        background: "#EEF2FF",
        text: "#312E81",
      },
      generationId: "mock-refine-123",
      createdAt: new Date().toISOString(),
      originalImageUrl: "https://example.com/mock-concept.png",
      refinementPrompt: "Make it more vibrant",
    },
    error: {
      status: 500,
      message: "Failed to refine concept",
    },
  },
};

/**
 * Configuration options for the mock API service
 */
export interface MockApiConfig {
  shouldFail?: boolean;
  responseDelay?: number;
  customResponses?: {
    generateConcept?: Partial<GenerationResponse>;
    refineConcept?: Partial<GenerationResponse>;
  };
}

/**
 * Default configuration
 */
const defaultConfig: MockApiConfig = {
  shouldFail: false,
  responseDelay: 500,
};

/**
 * Mock API service that simulates the concept visualizer backend
 */
export class MockApiService {
  private config: MockApiConfig;

  constructor(config: MockApiConfig = defaultConfig) {
    this.config = { ...defaultConfig, ...config };
  }

  /**
   * Simulates a delayed response
   */
  private async delay(): Promise<void> {
    return new Promise((resolve) =>
      setTimeout(resolve, this.config.responseDelay),
    );
  }

  /**
   * Simulates a response with the specified data or error
   */
  private async simulateResponse<T>(
    successData: T,
    errorData: ApiError,
  ): Promise<ApiResponse<T>> {
    await this.delay();

    if (this.config.shouldFail) {
      return { error: errorData, loading: false };
    }

    return { data: successData, loading: false };
  }

  /**
   * Simulates the concept generation endpoint
   */
  async generateConcept(
    request: PromptRequest,
  ): Promise<ApiResponse<GenerationResponse>> {
    // Validate inputs (similar to real implementation)
    if (!request.logoDescription || !request.themeDescription) {
      return {
        error: {
          status: 400,
          message: "Logo and theme descriptions are required",
        },
        loading: false,
      };
    }

    // Use custom response if provided or fallback to default mock
    const successResponse = {
      ...mockResponses.generateConcept.success,
      ...this.config.customResponses?.generateConcept,
    };

    return this.simulateResponse<GenerationResponse>(
      successResponse,
      mockResponses.generateConcept.error,
    );
  }

  /**
   * Simulates the concept refinement endpoint
   */
  async refineConcept(
    request: RefinementRequest,
  ): Promise<ApiResponse<GenerationResponse>> {
    // Validate inputs (similar to real implementation)
    if (!request.originalImageUrl || !request.refinementPrompt) {
      return {
        error: {
          status: 400,
          message: "Original image URL and refinement prompt are required",
        },
        loading: false,
      };
    }

    // Use custom response if provided or fallback to default mock
    const successResponse = {
      ...mockResponses.refineConcept.success,
      ...this.config.customResponses?.refineConcept,
      originalImageUrl: request.originalImageUrl,
      refinementPrompt: request.refinementPrompt,
    };

    return this.simulateResponse<GenerationResponse>(
      successResponse,
      mockResponses.refineConcept.error,
    );
  }

  /**
   * Configure the mock API service
   */
  configure(config: Partial<MockApiConfig>): void {
    this.config = { ...this.config, ...config };
  }
}

/**
 * Singleton instance of the mock API service
 */
export const mockApiService = new MockApiService();

/**
 * Creates a wrapped API response matching the useApi hook's return value
 */
export function createMockApiResponse<T>(
  data?: T,
  error?: ApiError,
): ApiResponse<T> {
  return {
    data,
    error,
    loading: false,
  };
}
