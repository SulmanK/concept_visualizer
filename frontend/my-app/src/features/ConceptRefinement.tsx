import React, { useState } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import { ConceptRefinementForm } from '../components/concept/ConceptRefinementForm';
import { ConceptResult } from '../components/concept/ConceptResult';
import { Card } from '../components/ui/Card';
import { Button } from '../components/ui/Button';
import { useConceptRefinement } from '../hooks/useConceptRefinement';

/**
 * Feature component for refining existing concepts
 */
export const ConceptRefinement: React.FC = () => {
  const { conceptId } = useParams<{ conceptId: string }>();
  const navigate = useNavigate();
  
  // In a real app, we would fetch the concept using the conceptId
  // For demo purposes, we'll use a mock concept
  const mockOriginalConcept = {
    imageUrl: 'https://placehold.co/800x800?text=Original+Concept',
    logoDescription: 'A modern logo for a tech company',
    themeDescription: 'Blue and purple futuristic theme',
  };
  
  const { 
    refineConcept, 
    resetRefinement, 
    status, 
    result, 
    error,
    isLoading
  } = useConceptRefinement();
  
  const [selectedColor, setSelectedColor] = useState<string | null>(null);
  
  const handleRefineConcept = (
    refinementPrompt: string,
    logoDescription: string,
    themeDescription: string,
    preserveAspects: string[]
  ) => {
    refineConcept(
      mockOriginalConcept.imageUrl,
      refinementPrompt,
      logoDescription || undefined,
      themeDescription || undefined,
      preserveAspects
    );
    setSelectedColor(null);
  };
  
  const handleReset = () => {
    resetRefinement();
    setSelectedColor(null);
  };
  
  const handleCancel = () => {
    navigate('/');
  };
  
  const handleColorSelect = (color: string) => {
    setSelectedColor(color);
    
    // Copy color to clipboard
    navigator.clipboard.writeText(color)
      .then(() => {
        console.log(`Copied ${color} to clipboard`);
      })
      .catch(err => {
        console.error('Could not copy text: ', err);
      });
  };
  
  // Loading skeleton for when we're fetching the original concept
  if (!mockOriginalConcept) {
    return (
      <Card isLoading={true} className="max-w-xl mx-auto">
        <div className="h-64"></div>
      </Card>
    );
  }
  
  return (
    <div className="space-y-8">
      <div className="text-center mb-8">
        <h1 className="text-3xl font-bold text-dark-900 mb-2">
          Refine Your Concept
        </h1>
        <p className="text-dark-600 max-w-2xl mx-auto">
          Provide instructions to refine your existing concept. You can update the logo or theme description, or specify what aspects to preserve.
        </p>
      </div>
      
      {!result && (
        <ConceptRefinementForm
          originalImageUrl={mockOriginalConcept.imageUrl}
          onSubmit={handleRefineConcept}
          status={status}
          error={error}
          onCancel={handleCancel}
          initialLogoDescription={mockOriginalConcept.logoDescription}
          initialThemeDescription={mockOriginalConcept.themeDescription}
        />
      )}
      
      {status === 'success' && result && (
        <div className="mt-8 pt-8 border-t border-dark-200">
          <div className="flex justify-between mb-4">
            <Button variant="outline" onClick={handleReset}>
              Refine Again
            </Button>
            <Button variant="primary" onClick={() => navigate('/')}>
              Create New Concept
            </Button>
          </div>
          
          {/* Comparison view */}
          <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
            {/* Original concept */}
            <div>
              <h2 className="text-xl font-semibold text-dark-800 mb-4">Original</h2>
              <div className="border border-dark-200 rounded-lg overflow-hidden shadow-sm">
                <img 
                  src={mockOriginalConcept.imageUrl} 
                  alt="Original concept" 
                  className="w-full h-auto"
                />
              </div>
            </div>
            
            {/* Refined concept */}
            <div>
              <h2 className="text-xl font-semibold text-dark-800 mb-4">Refined</h2>
              <div className="border border-dark-200 rounded-lg overflow-hidden shadow-sm">
                <img 
                  src={result.imageUrl} 
                  alt="Refined concept" 
                  className="w-full h-auto"
                />
              </div>
            </div>
          </div>
          
          {/* Result with details */}
          <div className="mt-8">
            <ConceptResult
              concept={result}
              onRefineRequest={handleReset}
              onColorSelect={(color) => handleColorSelect(color)}
            />
            
            {selectedColor && (
              <div className="mt-4 text-center text-sm text-dark-600">
                <span className="bg-dark-100 px-2 py-1 rounded font-mono">
                  {selectedColor}
                </span>
                <span className="ml-2">
                  copied to clipboard
                </span>
              </div>
            )}
          </div>
        </div>
      )}
    </div>
  );
}; 