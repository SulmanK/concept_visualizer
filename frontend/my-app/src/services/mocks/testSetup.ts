/**
 * Configuration and utilities for using mock services in tests.
 * This file provides helpers to setup and reset mock services between tests.
 */

import { mockApiService, MockApiConfig } from './mockApiService';

/**
 * Setup the mock API for a test.
 * This allows configuring the mock API service for specific test scenarios.
 */
export function setupMockApi(config: Partial<MockApiConfig> = {}): void {
  mockApiService.configure(config);
}

/**
 * Reset the mock API to its default state.
 * Call this in beforeEach to ensure tests don't affect each other.
 */
export function resetMockApi(): void {
  mockApiService.configure({
    shouldFail: false,
    responseDelay: 0, // Use 0 for fast tests
    customResponses: undefined,
  });
}

/**
 * Configure the mock API to simulate failures.
 */
export function mockApiFailure(): void {
  mockApiService.configure({
    shouldFail: true,
  });
}

/**
 * Configure the mock API with custom responses.
 */
export function mockApiCustomResponse(customResponses: MockApiConfig['customResponses']): void {
  mockApiService.configure({
    customResponses,
  });
}

/**
 * Setup mocks for all tests in a test file.
 * Call this in the describe block outside of any individual test.
 */
export function setupTestMocks(): void {
  beforeEach(() => {
    resetMockApi();
    jest.clearAllMocks();
  });
  
  afterEach(() => {
    jest.restoreAllMocks();
  });
} 