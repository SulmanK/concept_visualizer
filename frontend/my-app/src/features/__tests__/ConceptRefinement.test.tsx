import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { ConceptRefinement } from '../ConceptRefinement';
import * as useConceptRefinementModule from '../../hooks/useConceptRefinement';
import { vi } from 'vitest';
import { mockApiService } from '../../services/mocks/mockApiService';
import { setupMockApi, resetMockApi, mockApiFailure } from '../../services/mocks/testSetup';

// Mock useNavigate
const mockNavigate = vi.fn();
vi.mock('react-router-dom', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useNavigate: () => mockNavigate,
    useParams: () => ({ conceptId: 'mock-concept-123' })
  };
});

// Helper for renderWithRouter
const renderWithRouter = (ui: React.ReactElement) => {
  return render(
    <BrowserRouter>
      <Routes>
        <Route path="*" element={ui} />
      </Routes>
    </BrowserRouter>
  );
};

// Setup mockApiService for specific scenarios
const setupSuccessScenario = () => {
  setupMockApi({
    responseDelay: 0,
    customResponses: {
      refineConcept: {
        imageUrl: 'https://example.com/refined-concept.png',
        colorPalette: {
          primary: '#6366F1',
          secondary: '#A5B4FC',
          accent: '#E0E7FF',
          background: '#EEF2FF',
          text: '#312E81'
        },
        generationId: 'refined-concept-123',
        createdAt: new Date().toISOString(),
        originalImageUrl: 'https://example.com/original-concept.png',
        refinementPrompt: 'Make it more vibrant'
      }
    }
  });
};

describe('ConceptRefinement Component', () => {
  beforeEach(() => {
    resetMockApi();
    vi.clearAllMocks();
    
    // Restore the original implementation for all tests
    vi.restoreAllMocks();
    
    // Mock localStorage to provide a test concept
    const mockLocalStorage = {
      getItem: vi.fn().mockImplementation((key) => {
        if (key === 'savedConcepts') {
          return JSON.stringify([
            {
              imageUrl: 'https://example.com/original-concept.png',
              colorPalette: {
                primary: '#4F46E5',
                secondary: '#818CF8',
                accent: '#C4B5FD',
                background: '#F5F3FF',
                text: '#1E1B4B'
              },
              generationId: 'mock-concept-123',
              createdAt: new Date().toISOString()
            }
          ]);
        }
        return null;
      }),
      setItem: vi.fn()
    };
    
    Object.defineProperty(window, 'localStorage', {
      value: mockLocalStorage,
      writable: true
    });
  });
  
  test('renders form with original concept data', () => {
    renderWithRouter(<ConceptRefinement />);
    
    // Original concept should be displayed
    const originalImage = screen.getByAltText(/Original concept/i);
    expect(originalImage).toBeInTheDocument();
    expect(originalImage).toHaveAttribute('src', 'https://example.com/original-concept.png');
    
    // Form elements should be present
    expect(screen.getByLabelText(/Refinement Instructions/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Refine Concept/i })).toBeInTheDocument();
  });
  
  test('handles form submission for refinement', async () => {
    // Setup mock for successful refinement
    setupSuccessScenario();
    
    renderWithRouter(<ConceptRefinement />);
    
    // Fill out the refinement form
    const instructionsInput = screen.getByLabelText(/Refinement Instructions/i);
    fireEvent.change(instructionsInput, { target: { value: 'Make it more vibrant with brighter colors' } });
    
    // Check some preserve aspects options
    const colorCheckbox = screen.getByLabelText(/Preserve color scheme/i);
    fireEvent.click(colorCheckbox);
    
    // Submit the form
    const submitButton = screen.getByRole('button', { name: /Refine Concept/i });
    fireEvent.click(submitButton);
    
    // Should show loading state
    expect(screen.getByText(/Refining your concept.../i)).toBeInTheDocument();
    
    // Wait for the result to appear
    await waitFor(() => {
      expect(screen.getByAltText(/Refined concept/i)).toBeInTheDocument();
    });
    
    // Check that the refined image is displayed
    const refinedImage = screen.getByAltText(/Refined concept/i);
    expect(refinedImage).toBeInTheDocument();
    expect(refinedImage).toHaveAttribute('src', 'https://example.com/refined-concept.png');
    
    // Should show both original and refined concept
    expect(screen.getByText(/Original Concept/i)).toBeInTheDocument();
    expect(screen.getByText(/Refined Concept/i)).toBeInTheDocument();
  });
  
  test('displays validation errors', () => {
    renderWithRouter(<ConceptRefinement />);
    
    // Try to submit with empty instructions
    const submitButton = screen.getByRole('button', { name: /Refine Concept/i });
    fireEvent.click(submitButton);
    
    // Should show validation error
    expect(screen.getByText(/Please provide refinement instructions/i)).toBeInTheDocument();
  });
  
  test('handles API errors gracefully', async () => {
    // Mock API failure
    mockApiFailure();
    
    renderWithRouter(<ConceptRefinement />);
    
    // Fill out the refinement form
    const instructionsInput = screen.getByLabelText(/Refinement Instructions/i);
    fireEvent.change(instructionsInput, { target: { value: 'Make it more vibrant' } });
    
    // Submit the form
    const submitButton = screen.getByRole('button', { name: /Refine Concept/i });
    fireEvent.click(submitButton);
    
    // Wait for the error to appear
    await waitFor(() => {
      expect(screen.getByText(/Failed to refine concept/i)).toBeInTheDocument();
    });
    
    // Form should still be accessible
    expect(screen.getByRole('button', { name: /Refine Concept/i })).toBeInTheDocument();
  });
  
  test('navigates back to home when cancel is clicked', () => {
    renderWithRouter(<ConceptRefinement />);
    
    // Find and click the Cancel button
    const cancelButton = screen.getByRole('button', { name: /Cancel/i });
    fireEvent.click(cancelButton);
    
    // Should navigate back to home
    expect(mockNavigate).toHaveBeenCalledWith('/');
  });
  
  test('handles saving the refined concept', async () => {
    // Setup successful refinement scenario
    setupSuccessScenario();
    
    // Mock useConceptRefinement hook for success state
    vi.spyOn(useConceptRefinementModule, 'useConceptRefinement').mockImplementation(() => ({
      refineConcept: vi.fn(),
      status: 'success',
      result: {
        imageUrl: 'https://example.com/refined-concept.png',
        colorPalette: {
          primary: '#6366F1',
          secondary: '#A5B4FC',
          accent: '#E0E7FF',
          background: '#EEF2FF',
          text: '#312E81'
        },
        generationId: 'refined-concept-123',
        createdAt: new Date().toISOString(),
        originalImageUrl: 'https://example.com/original-concept.png',
        refinementPrompt: 'Make it more vibrant'
      },
      error: null,
      isLoading: false,
      clearError: vi.fn()
    }));
    
    renderWithRouter(<ConceptRefinement />);
    
    // Find and click the Save Refined Concept button
    const saveButton = screen.getByRole('button', { name: /Save Refined Concept/i });
    fireEvent.click(saveButton);
    
    // Should save to localStorage and navigate to home
    expect(window.localStorage.setItem).toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith('/');
  });
}); 