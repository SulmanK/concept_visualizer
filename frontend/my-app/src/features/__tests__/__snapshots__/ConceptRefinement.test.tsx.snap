// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConceptRefinement Component > matches snapshot in initial state 1`] = `
<div
  class="space-y-8"
>
  <div
    class="text-center mb-8"
  >
    <h1
      class="text-3xl font-bold text-dark-900 mb-2"
    >
      Refine Your Concept
    </h1>
    <p
      class="text-dark-600 max-w-2xl mx-auto"
    >
      Provide instructions to refine your existing concept. You can update the logo or theme description, or specify what aspects to preserve.
    </p>
  </div>
  <div
    class="card-gradient max-w-xl mx-auto"
  >
    <div
      class="px-4 py-3 sm:px-6 border-b border-indigo-200 bg-indigo-50/50"
    >
      <h2
        class="text-xl font-semibold text-dark-900"
      >
        Refine Concept
      </h2>
    </div>
    <div
      class="p-4 sm:p-6"
    >
      <form
        class="space-y-4"
      >
        <div
          class="flex justify-center mb-4"
        >
          <div
            class="w-40 h-40 border border-dark-200 rounded-lg overflow-hidden shadow-sm"
          >
            <img
              alt="Original concept"
              class="w-full h-full object-cover"
              src="https://placehold.co/800x800?text=Original+Concept"
            />
          </div>
        </div>
        <div>
          <div
            class="w-full"
          >
            <label
              class="label"
              for="textarea-nk6tknt3k"
            >
              Refinement Instructions
            </label>
            <textarea
              aria-describedby="textarea-nk6tknt3k-helper"
              aria-invalid="false"
              class="input resize-y  w-full"
              id="textarea-nk6tknt3k"
              placeholder="Describe how you want to refine this concept..."
              rows="3"
            />
            <p
              class="helper-text"
              id="textarea-nk6tknt3k-helper"
            >
              Be specific about what you want to change
            </p>
          </div>
        </div>
        <div>
          <div
            class="w-full"
          >
            <label
              class="label"
              for="textarea-00ku13t95"
            >
              Updated Logo Description (Optional)
            </label>
            <textarea
              aria-describedby="textarea-00ku13t95-helper"
              aria-invalid="false"
              class="input resize-y  w-full"
              id="textarea-00ku13t95"
              placeholder="Update the logo description or leave as is..."
              rows="2"
            >
              A modern logo for a tech company
            </textarea>
            <p
              class="helper-text"
              id="textarea-00ku13t95-helper"
            >
              Leave empty to keep original description
            </p>
          </div>
        </div>
        <div>
          <div
            class="w-full"
          >
            <label
              class="label"
              for="textarea-gl3759cij"
            >
              Updated Theme Description (Optional)
            </label>
            <textarea
              aria-describedby="textarea-gl3759cij-helper"
              aria-invalid="false"
              class="input resize-y  w-full"
              id="textarea-gl3759cij"
              placeholder="Update the theme description or leave as is..."
              rows="2"
            >
              Blue and purple futuristic theme
            </textarea>
            <p
              class="helper-text"
              id="textarea-gl3759cij-helper"
            >
              Leave empty to keep original description
            </p>
          </div>
        </div>
        <div>
          <p
            class="label mb-2"
          >
            Preserve Aspects (Optional)
          </p>
          <div
            class="flex flex-wrap gap-3"
          >
            <label
              class="flex items-center space-x-2 cursor-pointer"
            >
              <input
                class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
                type="checkbox"
              />
              <span
                class="text-sm text-dark-700"
              >
                Layout
              </span>
            </label>
            <label
              class="flex items-center space-x-2 cursor-pointer"
            >
              <input
                class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
                type="checkbox"
              />
              <span
                class="text-sm text-dark-700"
              >
                Colors
              </span>
            </label>
            <label
              class="flex items-center space-x-2 cursor-pointer"
            >
              <input
                class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
                type="checkbox"
              />
              <span
                class="text-sm text-dark-700"
              >
                Style
              </span>
            </label>
            <label
              class="flex items-center space-x-2 cursor-pointer"
            >
              <input
                class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
                type="checkbox"
              />
              <span
                class="text-sm text-dark-700"
              >
                Symbols/Icons
              </span>
            </label>
            <label
              class="flex items-center space-x-2 cursor-pointer"
            >
              <input
                class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
                type="checkbox"
              />
              <span
                class="text-sm text-dark-700"
              >
                Proportions
              </span>
            </label>
          </div>
        </div>
        <div
          class="flex justify-end space-x-3 pt-2"
        >
          <button
            class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed border border-indigo-300 text-indigo-700 bg-white hover:bg-indigo-50 hover:text-primary-dark text-sm px-4 py-2 rounded-lg "
            type="button"
          >
            Cancel
          </button>
          <button
            class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed bg-gradient-primary text-white shadow-modern hover:shadow-modern-hover hover:brightness-105 text-sm px-4 py-2 rounded-lg "
            type="submit"
          >
            Refine Concept
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
